cobalt/js_profiler/profiler.cc

```diff
@@ -0,0 +1,147 @@
+                              options.max_buffer_size(), sample_interval_us),
+      settings);
+
+  if (status == v8::CpuProfilingStatus::kAlreadyStarted) {
+    web::DOMException::Raise(web::DOMException::kInvalidStateErr,
+                             "Profiler Already started", exception_state);
+  } else if (status == v8::CpuProfilingStatus::kErrorTooManyProfilers) {
+    web::DOMException::Raise(web::DOMException::kInvalidStateErr,
+                             "Too Many Profilers", exception_state);
+  }
+}
+
+v8::CpuProfilingStatus Profiler::ImplProfilingStart(
+    std::string profiler_id, v8::CpuProfilingOptions options,
+    script::EnvironmentSettings* settings) {
+  auto isolate = web::get_isolate(settings);
+  cpu_profiler_ = v8::CpuProfiler::New(isolate);
+  cpu_profiler_->SetSamplingInterval(kBaseSampleIntervalMs *
+                                     base::Time::kMicrosecondsPerMillisecond);
+  return cpu_profiler_->StartProfiling(
+      toV8String(isolate, profiler_id), options,
+      std::make_unique<ProfilerMaxSamplesDelegate>(this));
+}
+
+std::string Profiler::nextProfileId() {
+  s_lastProfileId++;
+  return "cobalt::profiler[" + std::to_string(s_lastProfileId) + "]";
+}
+
+void Profiler::PerformStop(
+    script::EnvironmentSettings* environment_settings,
+    std::unique_ptr<script::ValuePromiseWrappable::Reference> promise_reference,
+    base::TimeTicks time_origin, std::string profiler_id) {
+  auto isolate = web::get_isolate(environment_settings);
+  auto profile =
+      cpu_profiler_->StopProfiling(toV8String(isolate, profiler_id_));
+  auto trace = ProfilerTraceBuilder::FromProfile(profile, time_origin_);
+  scoped_refptr<ProfilerTraceWrapper> result(new ProfilerTraceWrapper(trace));
+  promise_reference->value().Resolve(result);
+}
+
+Profiler::ProfilerTracePromise Profiler::Stop(
+    script::EnvironmentSettings* environment_settings) {
+  script::HandlePromiseWrappable promise =
+      web::get_script_value_factory(environment_settings)
+          ->CreateInterfacePromise<scoped_refptr<ProfilerTraceWrapper>>();
+  if (!stopped()) {
+    stopped_ = true;
+    auto* global_wrappable = web::get_global_wrappable(environment_settings);
+    auto* context = web::get_context(environment_settings);
+    std::unique_ptr<script::ValuePromiseWrappable::Reference> promise_reference(
+        new script::ValuePromiseWrappable::Reference(global_wrappable,
+                                                     promise));
+
+    context->message_loop()->task_runner()->PostTask(
+        FROM_HERE,
+        base::BindOnce(&Profiler::PerformStop, base::Unretained(this),
+                       environment_settings, std::move(promise_reference),
+                       std::move(time_origin_), std::move(profiler_id_)));
+  } else {
+    promise->Reject(new web::DOMException(web::DOMException::kInvalidStateErr,
+                                          "Profiler already stopped."));
```

Comment:
Did you test this actually works? I think this may have to be done after returning the promise to JS (using a Posttask)) for JS to detect the rejection.

Reply:
I just triple-checked and this works. We're just rejecting the JS promise right away instead of scheduling a future-task. There are other examples in the cobalt codebase that follow this pattern (like in cache.cc).